### PR TITLE
fix: break inside async closure has incorrect span for enclosing closure

### LINKED ISSUE #124496: `break` inside async closure has incorrect span for enclosing closure
### Code

```rust
#![feature(async_closure)]

fn main() {
    let _ = || break;
    let _ = async || break;
}
```

### Current output

```console
error[E0267]: `break` inside of a closure
 --> src/main.rs:4:16
  |
4 |     let _ = || break;
  |             -- ^^^^^ cannot `break` inside of a closure
  |             |
  |             enclosing closure

error[E0267]: `break` inside of a closure
 --> src/main.rs:5:22
  |
5 |     let _ = async || break;
  |                      ^^^^^
  |                      |
  |                      cannot `break` inside of a closure
  |                      enclosing closure
```

### Desired output

```console
...

error[E0267]: `break` inside of a closure
 --> src/main.rs:5:13
  |
5 |     let _ = async || break;
  |             -------- ^^^^^
  |             |        cannot `break` inside of a closure
  |             |
  |             enclosing closure
```

### Rust Version

```Shell
rustc 1.79.0-nightly (aed2187d5 2024-04-27)
binary: rustc
commit-hash: aed2187d53b8789e3a37f50ae36f894a2a679077
commit-date: 2024-04-27
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.4
```

### PR DESCRIPTION
**Pull request by linyihai**
Fixes #124496

### CODE REVIEW

File: compiler/rustc_passes/src/loops.rs [hunk 1]
```diff
@@ -91,6 +91,10 @@ impl<'a, 'hir> Visitor<'hir> for CheckLoopVisitor<'a, 'hir> {
             }) => {
                 let cx = match kind {
                     hir::ClosureKind::Coroutine(hir::CoroutineKind::Desugared(kind, source)) => {
+                        let fn_decl_span = match self.cx {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/loops.rs:94]: This seems like a bit too specific of a case here. I think we should just fix the `fn_decl_span` for the `hir::Closure` from an async closure.

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Pls look into fixing this in a more general way

**Comment by linyihai** [comment]
> Pls look into fixing this in a more general way

Hope my latest submit general enough, I haven't see other code modules, such as the hir: : CoroutineKind: : Desugared

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Sorry, I wasn't clear when I meant that this should be made more general. I think this needs to be fixed when the `hir::Closure` is *constructed*, somewhere in `rustc_ast_lowering`.

See https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_ast_lowering/expr.rs.html#1056-1122 or somewhere nearby.

**Comment by linyihai** [comment]
I think it's ready to review. Would you @compiler-errors  leave some feadback? I will be very appreciate for it

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**